pilot/platform/cloudfoundry/servicediscovery.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudfoundry
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+
+	copilotapi "code.cloudfoundry.org/copilot/api"
+
+	"istio.io/istio/pilot/model"
+)
+
+type CopilotClient interface {
+	copilotapi.IstioCopilotClient
+}
+
+const CF_APP_PORT = 8080
+
+type ServiceDiscovery struct {
+	Client CopilotClient
+}
+
+// Services list declarations of all services in the system
+func (sd *ServiceDiscovery) Services() ([]*model.Service, error) {
+	resp, err := sd.Client.Routes(context.Background(), new(copilotapi.RoutesRequest))
+	if err != nil {
+		return nil, fmt.Errorf("getting services: %s", err)
+	}
+	services := []*model.Service{}
+
+	for hostname, _ := range resp.Backends {
+		services = append(services, newService(hostname))
+	}
```

Comment:
since we know the length of the output array, we could do:
`services := make([]*model.Service, 0, len(resp.Backends))`